### PR TITLE
ParsePortSpec: handle IPv6 addresses

### PR DESCRIPTION
**Pull request by stapelberg**
split out of https://github.com/docker/docker/pull/20315
in order to fix https://github.com/docker/docker/issues/11518

Signed-off-by: Michael Stapelberg <stapelberg@google.com>

### CODE REVIEW

**Comment by lenovouser** [comment]
Can this be merged? This is kind of essential as you can't bind containers to IPv6-Addresses at the moment. @LK4D4 you seem the most active contributor from the Docker team. Can you review this?

File: nat/nat.go [hunk 1]
```diff
@@ -173,9 +173,31 @@ func ParsePortSpec(rawPort string) ([]PortMapping, error) {
...
+			return nil, fmt.Errorf("got %d parts, wanted 2: %q", len(nonIPParts), rawPort[ipEnd+2:])
+		}
+		parts["hostPort"] = nonIPParts[0]
+		parts["containerPort"] = nonIPParts[1]
+	} else {
+		var err error
+		parts, err = PartParser(portSpecTemplate, rawPort)
+		if err != nil {
+			return nil, err
+		}
```
Comment (dnephin) [inline nat/nat.go]: I think there is an easier way to handle this. `PartParser()` is an unfortunate abstraction. It's a generic way of escaping `:` separated strings with no way to escape `:` within the parts, and which is only used in a single place.  

This code effectively introduces a new `PartParser()` that supports escaping colons using square brackets, but doesn't take advantage of the built-in functions in `net` and net/url`.

I think we should fix this by deprecating `PartParser()`, and replacing it with a new function that just parses the port spec into the appropriate parts using the available built-in functions.

**Comment by dnephin** [comment]
We should add a test case as well, once we work out the implementation.

**Comment by stapelberg** [comment]
@dnephin I don’t have strong opinions on how the code should look like. Would you want to add a test and refactor it so that it corresponds to your ideals? :)

**Comment by dnephin** [comment]
I opened #19 with those changes

**Pull request closed without merging**